### DISCUSSION TITLE
Implementing custom application protocols

### DISCUSSION THREAD

**Discussion by pieterdd** [libp2p/rust-libp2p#3764] [Q&A] [answered]
Hi there! Newbie question. I've been playing around with the examples and I'm liking rust-libp2p so far. I combined a few NetworkBehaviours and I fairly quickly had a working program that finds peers through mDNS, connects to them via TCP with asymmetric encryption and exchanges identification info. It was a breeze.

Now I'd like to go beyond textbook examples. If I'm not mistaken, every example in the repo involves piecing together existing NetworkBehaviours into a composite data structure. What I'm looking for is some guidance on how to implement application-specific logic.

Suppose I want both peers to be able to ask for the other peer's time, and respond to such requests if they occur. Peers should only send requests and responses if both are running `/myprotocol/0.0.1`. If the `myprotocol` protocol is not available or the version mismatches, some kind of error handling should exist to provide feedback towards the user.

Is this a use case that one of the off-the-shelf NetworkBehaviours can be adapted to? Should I roll my own NetworkBehaviour and corresponding ConnectionHandler? Should I hook any application-specific logic to the `SwarmEvent::ConnectionEstablished` handler in the `swarm.select_next_some()` loop?

**Comment by thomaseizinger** [comment, marked as answer]
If your protocol is request-response style, you can use `libp2p-request-response`.

If it is not, you will have to implement the following traits on various structs:

- `ConnectionHandler` to do the state management per connection
- `NetworkBehaviour` to do the state management across your entire protocol

For a simple protocol such as answering the current time, that state is available "inline" with the request, meaning you can directly answer it in the stream upgrade (look at `InboundUpgrade` trait and how it is used).

However, for more complex state management, you want to use `ReadyUpgrade`, pass the stream to the `ConnectionHandler` and read and write to it there.

Have a look at how these traits are used. Unfortunately, there are many ways to solve this problem at the moment so there isn't one clear example to work from.

I'd say `libp2p-perf` is a relatively "modern" one.

**Reply by pieterdd** [reply]
Those are some helpful pointers, thanks!

**Reply by pieterdd** [reply]
I was looking to define a RequestMessage and ResponseMessage type to send commands/replies over the wire. If I understand the documentation well I should implement a `Codec` trait. I find myself trying to implement a codec that serializes/deserializes the request/response types to a sequence of bytes using serde, but that sounds like a very generic problem that has been solved before. Is there by any chance an existing `Codec` that lets you define an enum for the request messages and one for the response messages?

**Reply by thomaseizinger** [reply]
Not in our library but if it is a repeated problem in your app, it should be easy enough to build that abstraction yourself :)

**Reply by thomaseizinger** [reply]
Work is now being done it add this as a first-class citizen to our library: https://github.com/libp2p/rust-libp2p/pull/3952
